src/Dictionaries/TrieDictionary.cpp

```diff
@@ -331,13 +480,136 @@ void TrieDictionary::loadData()
             {
                 const auto & attribute_column = *attribute_column_ptrs[attribute_idx];
                 auto & attribute = attributes[attribute_idx];
-                setAttributeValue(attribute, key_column->getDataAt(row_idx), attribute_column[row_idx]);
+
+                setAttributeValue(attribute, attribute_column[row]);
+            }
+
+            size_t row_number = ip_records.size();
+
+            std::string addr_str(key_column->getDataAt(row).toString());
+            size_t pos = addr_str.find('/');
+            if (pos != std::string::npos)
+            {
+                IPAddress addr(addr_str.substr(0, pos));
+                has_ipv6 = has_ipv6 || (addr.family() == Poco::Net::IPAddress::IPv6);
+
+                UInt8 prefix = std::stoi(addr_str.substr(pos + 1), nullptr, 10);
+
+                addr = addr & IPAddress(prefix, addr.family());
+                ip_records.emplace_back(addr, prefix, row_number);
+            }
+            else
+            {
+                IPAddress addr(addr_str);
+                has_ipv6 = has_ipv6 || (addr.family() == Poco::Net::IPAddress::IPv6);
+                UInt8 prefix = addr.length() * 8;
+                ip_records.emplace_back(addr, prefix, row_number);
             }
         }
     }
 
     stream->readSuffix();
 
+    if (has_ipv6)
+    {
+        auto deleted_count = sort_and_unique(ip_records,
+            [](const auto & record_a, const auto & record_b)
+            {
+                uint8_t a_buf[IPV6_BINARY_LENGTH];
+                uint8_t b_buf[IPV6_BINARY_LENGTH];
+
+                auto cmpres = memcmp16(record_a.asIPv6Binary(a_buf), record_b.asIPv6Binary(b_buf));
+
+                if (cmpres == 0)
+                    return compareTo(record_a.prefixIPv6(), record_b.prefixIPv6());
+                return cmpres;
+            });
+        if (deleted_count > 0)
+            LOG_WARNING(logger, "removing {} non-unique subnets from input", deleted_count);
+
+        auto & ipv6_col = ip_column.emplace<IPv6Container>();
+        ipv6_col.resize_fill(IPV6_BINARY_LENGTH * ip_records.size());
+
+        for (const auto & record : ip_records)
+        {
+            auto ip_array = IPv6ToBinary(record.addr);
+
+            size_t i = row_idx.size();
+            memcpySmallAllowReadWriteOverflow15(&ipv6_col[i * IPV6_BINARY_LENGTH],
```

Comment:
I think we can update `IPv6ToBinary` a little bit to write result inplace (to avoid extra copy).
Copy is not expensive here, though  ...

Reply:
I create additional function `IPv6ToRawBinary`